Write create_translation, get_story_translation tests in test_story_s…

jennifer-lu opened the pull request:
…ervice.py

## Notion ticket link

<!-- Please replace with your ticket's URL -->

[Write create_translation, get_story_translation tests in test_story_service.py](https://www.notion.so/uwblueprintexecs/Write-create_translation-get_story_translation-tests-in-test_story_service-py-d87c038dee0f4b1f9f04b7c22f3e9c32)

<!-- Give a quick summary of the implementation details, provide design justifications if necessary -->

## Implementation description

- Implemented tests for create_translation and get_story_translation
- In test_get_story_translation, did not assert `resp["translator_name"]` or `resp["reviewer_name"]` because they are not returned in `get_story_translation`, but rather `get_story_translations`

<!-- What should the reviewer do to verify your changes? Describe expected results and include screenshots when appropriate -->

## Steps to test

1. Do tests pass?

<!-- Draw attention to the substantial parts of your PR or anything you'd like a second opinion on -->

## What should reviewers focus on?

- Is code readable?
- Are coding practices/decisions ok (please see comments)?

## Checklist

- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] For backend changes, I have run the appropriate linters: `docker exec -it planet-read_py-backend_1 /bin/bash -c "black . && isort --profile black ."` and I have generated new migrations: `flask db migrate -m "<your message>"`
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

jennifer-lu commented on backend/python/app/tests/services/test_story_service.py:
```diff
@@ -67,9 +72,44 @@ def test_get_stories_available_for_translation():
 
+    get_story_translation_resp = services["story"].get_story_translation(story_translation_obj.id)
+    content_objs = StoryContent.query.filter_by(story_id = story_resp.id).all()
+    assert len(content_objs) == get_story_translation_resp["num_content_lines"]
```
Is this the correct way to compare content lines?

gaoxk replied:
this is a good check! additionally you should check the actual lines on the response body, eg
`    assert len(content_objs) == len(get_story_translation_resp["translationContents"])`

jennifer-lu commented on backend/python/app/tests/services/test_story_service.py:
```diff
@@ -79,9 +119,65 @@ def test_create_translation_raises_error_if_translation_content_commit_fails_and
+            "translation_content": story_translation_content_obj_2.translation_content,
+            "status": story_translation_content_obj_2.status,
+        },
+    ]
```
Is there a better way to do this? I couldn't get it to work with `translation_contents = [story_translation_content_obj_1, story_translation_content_obj_2]`.

jennifer-lu commented on backend/python/app/tests/services/test_story_service.py:
```diff
@@ -79,9 +119,65 @@ def test_create_translation_raises_error_if_translation_content_commit_fails_and
+    assert resp["num_translated_lines"] == services["story"]._get_num_translated_lines(translation_contents)
+    assert resp["num_approved_lines"] == services["story"]._get_num_approved_lines(translation_contents)
+    assert resp["num_content_lines"] == len(translation_contents)
 
```
Didn't assert `resp["translator_name"]` or `resp["reviewer_name"]` because they aren't returned in `get_story_translation`, but rather `get_story_translations`.

jennifer-lu commented:
It seems that I made a new empty utils.sh file, should I delete it? Thanks!

gaoxk commented on backend/python/app/tests/services/test_story_service.py:
```diff
@@ -79,9 +137,103 @@ def test_create_translation_raises_error_if_translation_content_commit_fails_and
+
+    resp = services["story"].get_story_translation(story_translation_obj.id)
+
+    assert resp["id"] == story_translation_obj.id
```
thoughts on making something like `assert_story_equals_model` from story_helpers in story_translation_helpers? you can move all these asserts into the new function

gaoxk commented on backend/python/app/tests/services/test_story_service.py:
```diff
@@ -67,8 +73,60 @@ def test_get_stories_available_for_translation():
-def test_create_translation():
-    pass
+def test_create_translation(app, db, services):
+    alphanumeric = "abcdefghijklmnopqrstuvwxyz0123456789"
```
Im seeing repeated code for generating translator, reviewer, and story. thoughts on making something a helper function to handle this? like `create_story_translation` in https://github.com/uwblueprint/planet-read/pull/183/files

gaoxk commented on backend/python/app/tests/services/test_story_service.py:
```diff
@@ -67,8 +73,60 @@ def test_get_stories_available_for_translation():
+def test_create_translation(app, db, services):
+    alphanumeric = "abcdefghijklmnopqrstuvwxyz0123456789"
+
+    translator_authId = "".join(random.choice(alphanumeric) for i in range(10))
```
imo i dont think we have to go through the trouble of making randomly generated ids but I also think it's fine that the code for making random ids is here

gaoxk reviewed (changes requested):
looking good! :D 

>It seems that I made a new empty utils.sh file, should I delete it? Thanks!

Delete it - you should be able to do so with `git rm utils.sh`

jennifer-lu commented:
> 

I deleted the empty utils.sh file, but then it deleted the original, so I reverted that commit. I'm pretty confused about what happened. I'll leave the utils.sh file for now.

gaoxk commented on backend/python/app/tests/helpers/story_translation_helpers.py:
```diff
@@ -0,0 +1,141 @@
+            status="DEFAULT",
+        ),
+    )
+    story_translation_contents = [
```
hmmm this seems repetitive information and has space for human error, what about 
`story_translation_contents = [story_translation_content_1.__dict__, story_translation_content_2.__dict__]`

jennifer-lu replied:
Thanks for this nit (and the reviews)! I had to filter some keys because `story_translation_response["translation_contents"]` returns fewer keys per `translation_content` than `story_translation_content.to_dict()` does.

gaoxk commented on backend/python/app/tests/services/test_story_service.py:
```diff
@@ -79,9 +119,35 @@ def test_create_translation_raises_error_if_translation_content_commit_fails_and
+        story_translation_contents
+    )
+
+    assert_story_translation_equals_model(
```
nice! very clean

gaoxk reviewed (approved):
feel free to merge after the nit!
